controllers/install/binoculars_controller.go

```diff
@@ -112,104 +112,114 @@ func (r *BinocularsReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 	mutateFn := func() error { return nil }
 
-	if components.ServiceAccount != nil {
+	_, err = upsertComponents(ctx, components, r.Client, mutateFn, controllerutil.CreateOrUpdate)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Successfully reconciled Binoculars object", "durationMilis", time.Since(started).Milliseconds())
+
+	return ctrl.Result{}, nil
+}
+
+type BinocularsComponents struct {
+	ClusterRole        *rbacv1.ClusterRole
+	ClusterRoleBinding *rbacv1.ClusterRoleBinding
+	Ingress            *networking.Ingress
+	IngressRest        *networking.Ingress
+	Deployment         *appsv1.Deployment
+	Service            *corev1.Service
+	ServiceAccount     *corev1.ServiceAccount
+	Secret             *corev1.Secret
+}
+
+func upsertComponents(ctx context.Context, bc *BinocularsComponents, r client.Client, mutateFn func() error, createOrUpdate func(ctx context.Context, c client.Client, obj client.Object, f controllerutil.MutateFn) (controllerutil.OperationResult, error)) (ctrl.Result, error) {
```

Comment:
can this be reformatted on some sort of multi-line way?

Reply:
I made it a struct that you can pass in.